### DISCUSSION TITLE
ECS

### DISCUSSION THREAD

**Discussion by ghost** [AxiomGames/Axiom#18] [General]
About ECS, even though there are really cool ECSs out there, in my opinion it's a vital part of the game engine, that maybe we should have total control over it.

- First question: Full ECS or just EC? 

Just to clarify: a truly ECS make the separation between logic and data, there is no "ScriptComponent" with a virtual OnUpdate() inside, ike Unity. there are a nice explanation here: https://github.com/SanderMertens/ecs-faq#how-is-ecs-different-from-entity-component-frameworks.

I think one of the main advantages of having a truly ECS is: **Performance**, because of the cache hit.
But you can have other advantages, the level of decoupling can be insane.

ps: I often see people using entt internally in the engine and exposing a "ScriptComponent" to the game developer, (I think Hazel do that), it could have some advantages, especially for people that are used to make games with Unity, but mixing these concepts could waste the performance advantages of using a ECS due to a lot of cache misses that will happen

-  So, ECS or a entity-component-framework(EC) ? 

In my opition, going with the **full ECS** is the right way, even Unity is moving to this direction (see Unity DOTS).

In some tests I've done, I can easily iterate over 10 million entities in less than 10ms just using a single-threaded application.
if we add multithreading (which can be done using some type of command buffer to defer the changes) we can have a even better result

But now that I'm using a ECS for 2 years, I can see some drawbacks, the first one that come to my mind is related to "Syncing a data-oriented ECS with a stateful external system", got this title from here: https://ruby0x1.github.io/machinery_blog_archive/post/syncing-a-data-oriented-ecs/index.html **TL;DL:** if someone change the transform of a entity, we should notify all systems that could cache this value, like inside physics or the renderer that could to have a BVH. using classic OOP, we can use the observer pattern. using true DOP components, that's not possible. 

(I have some possible solutions to this issue, but this post is becoming too long )

**Comment by SamCZ** [comment]
I would totally like go for full ECS, I wrote my own and we can totally use that, but its just the EC in that, we still need to find a good way to use systems, since I find them kind of a pain to use in gameplay.

**Comment by ghost** [comment]
if you already know the EC, I don't think it's a pain to use in gameplay

for instance, using EC you will do something like that:
```
class MoveScript : ScriptComponent {
	
	public float speed;
	private TransformComponent transform;

	public override void Init() {
		transform = GetComponent<TransformComponent>();
	}

	public override void Update() {
		if (InputManager.IsKeyPressed(Keys.W)) {
			transform.position.y -= speed
		}
	}	
};
```

using ECS it's basically the same thing, you just separate the data and the logic:

```
struct MoveComponent {
	float speed;
}

void UpdateSystem() {
	Query<TransformComponent, MoveComponent> query; //get query or view(using entt naming)
	query.each([](TransformComponent& transform, MoveComponent& moveComponent){
		if (InputManager.IsKeyPressed(Keys.W)) {
			transform.position.y -= moveComponent.speed;
		}
	});
}
```

**Reply by benanil** [reply]
Isn't using for loops faster than passing function? because compiler can optimize&vectorize the code for you
but if we pass the function as an argument compiler wouldn't be able to optimize and it will just call the function
I mean it is still memory efficient and we hit the cache line better with ECS but. Is there any way we can create an ECS system that we can iterate over entities with for/while loop.
https://www.youtube.com/watch?v=rX0ItVEVjHc

**Reply by SamCZ** [reply]
In my ECS I have created iterator that just goes through the component memory and its quite efficient.

**Reply by ghost** [reply]
yep, we can do like entt:

```
for(auto [entity, pos, vel]: view.each()) {
}
```

**Reply by SamCZ** [reply]
But what I found while developing the other engine is that you want option to have same component type in entity multiple times. for example multiple scripts or lights or colliders.

**Reply by benanil** [reply]
Paulo that's nice. I would love to see generated code.

**Reply by ghost** [reply]
@SamCZ having more than one component of the same type in the same entity it's a common discussion about ECS. but I never saw a ECS that do that.

But why should you need to have multiple lights or colliders in the same entity?

**Reply by SamCZ** [reply]
What if you have mesh and you cannot fit only one box collider on it ?

**Reply by SamCZ** [reply]
If you look at Unity they have option to have multiple types of same components.

**Reply by SamCZ** [reply]
I found it pretty limiting in my other engine when I have option to have only one type of component

**Reply by SamCZ** [reply]
But yeah having multiple of same component types of more complicated

**Reply by ghost** [reply]
it will make the ECS a way more complex, not only to implement, but to use as well.  IDK I think I still prefer to create a custom component to handle this things, something like: "MultiBoxColliderComponent", or inside the "BoxColliderComponent" add the option to have multiple colliders

**Reply by SamCZ** [reply]
Yeah I though about it too, but it will remove the continuous memory, since you will need to have some kind of array with optional size inside.

**Reply by SamCZ** [reply]
@PauloMarangoni

**Reply by ghost** [reply]
and you are right, Unity allows multiple components of the same type 
https://docs.unity3d.com/ScriptReference/GameObject.GetComponents.html

but if we used this thing in the Update function it could require at least one heap allocation per entity/frame
```
List<HingeJoint> hingeJoints = new List<HingeJoint>();
GetComponents(hingeJoints);
foreach (HingeJoint joint in hingeJoints)
    joint.useSpring = false;
```

**Reply by SamCZ** [reply]
@PauloMarangoni You can again just use iterators and dont need to allocator new vector or smth.

**Reply by ghost** [reply]
@SamCZ yeah you are right about the memory, but it's hard to be perfect here, my only concern about having multiple components of the same type is adding a unnecessary complexity just to handle some exceptions. if we can find a way to do it without adding extra completixity, why not?

**Reply by ghost** [reply]
in my iterator example I have two ideas that maybe could work

```
for(auto [entity, collider]: view.each()) {
}
```
we can:
     1) execute twice for the same entity if we have multiple components.
     2) make the collider a iterator too. (all components now should be treated as iterators? it will make more complex)

**Reply by benanil** [reply]
ECS is not always best solution depending on the situation we can use different design patterns as such

```cpp
struct EnemySystem
{
    float* positionX;
    float* positionY;
    float* positionZ;
    float* health;
    float* damage;
};
```
this type of structure of arrays system might be usefull in some cases

**Reply by benanil** [reply]
so maybe for gameplay code we can use object orianted programming, I mean instead of sticking with same pattern we can mix the design patterns.

**Reply by SamCZ** [reply]
We can have it like this:
```cpp
for (Entity ent : view)
{
 TransformComponent& trans = ent.GetFirstComponent<TransformComponent>();
 for (BoxColliderComponent& boxCollider, ent.GetComponentsOfSameType<BoxColliderComponent>)
 {
  ....
 }
}
```
`GetFirstComponent()` will just access the memory based on the entity ID stored inside the `Entity` struct
`GetComponentsOfSameType<>()` Will return iterators to the same components.

**Reply by SamCZ** [reply]
> ECS is not always best solution depending on the situation we can use different design patterns as such
> 
> ```c++
> struct EnemySystem
> {
>     float* positionX;
>     float* positionY;
>     float* positionZ;
>     float* health;
>     float* damage;
> };
> ```
> 
> this type of structure of array system might be usefull in some cases

this is horrible and it will be worst managable code

**Reply by benanil** [reply]
nope it is not for Mike Acton, programmer of Unity Dots and insomniac games, (the game that I loved optimization Spider-Man)

**Reply by benanil** [reply]
if you look at the youtube video that I send you will se what I mean, and you will see it in Data orianted programming books
it is hard to manage as you said but not horible

**Reply by SamCZ** [reply]
You know that the ESC we walk about is data oriented ? We just expose nice api to the user

**Reply by SamCZ** [reply]
@benanil

**Reply by benanil** [reply]
I didn't send that data structure as an alternative of ECS. I said we can make different systems for different things
ECS can't solve every problem

**Reply by benanil** [reply]
the reason that I send the structure above is they've used simd instructions instead of if statements
and performed mathematical operations with simd and it was way faster because memory layout was very convenient to that problem(it was an cppcon talk but I can't find rn)

**Reply by benanil** [reply]
for example player is single entity we don't need ECS for that. there is no reason to use ECS

**Reply by ghost** [reply]
> ECS is not always best solution depending on the situation we can use different design patterns as such
> 
> ```c++
> struct EnemySystem
> {
>     float* positionX;
>     float* positionY;
>     float* positionZ;
>     float* health;
>     float* damage;
> };
> ```
> 
> this type of structure of arrays system might be usefull in some cases

yes, you are right, we should not use ECS to solve all problems, for instance, if we need to send data to the GPU we can't rely on the ECS structure.

but at the moment, I think ECS is a good trade-off of having a very fast but also simple and clean way to solve most of the problems related to game logic.

**Reply by ghost** [reply]
> for example player is single entity we don't need ECS for that. there is no reason to use ECS

that's the beauty of having a true ECS, you don't need it, if you don't want.
The code is not attached to a entity like Unity.

**Reply by ghost** [reply]
> We can have it like this:
> 
> ```c++
> for (Entity ent : view)
> {
>  TransformComponent& trans = ent.GetFirstComponent<TransformComponent>();
>  for (BoxColliderComponent& boxCollider, ent.GetComponentsOfSameType<BoxColliderComponent>)
>  {
>   ....
>  }
> }
> ```
> 
> `GetFirstComponent()` will just access the memory based on the entity ID stored inside the `Entity` struct `GetComponentsOfSameType<>()` Will return iterators to the same components.

@SamCZ I tend to avoid using "gets" in the systems, even if you are getting the component from the view, there is some performance penalty for searching the component.

using callbacks or range-for, we can do this "search" in compile time, we just need to respect the order in the view.

obviously I have in my interface, it could be handy but that's not the default way of use the ECS.

**Reply by ghost** [reply]
@SamCZ it could work too, not sure how hard is to implement
ps:  and maybe we will lost the continuous memory the same way

```
View<Transform, Iterator<BoxColliderComponent>> view; //get the view
for(auto [entity, transform, boxIterator]: view.each()) {	
	for(auto& boxCollider: boxIterator) {

	}
}
```

**Reply by benanil** [reply]
> > for example player is single entity we don't need ECS for that. there is no reason to use ECS
> 
> that's the beauty of having a true ECS, you don't need it, if you don't want. The code is not attached to a entity like Unity.

Actually what unity doing was 'GameObject component system' but they have improved that system and they work on DOTS with Mike Acton

**Reply by ghost** [reply]
DOTS is a ECS, and that's exactly the same thing we are talking here
